File: kotlin/src/test/java/cn/staynoob/trap/java/basic/TypeInferenceSpec.java
package cn.staynoob.trap.java.basic;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

public class TypeInferenceSpec {

    static abstract class AbstractBase<T> {

        AbstractBase(T value) {
        }

    }

    // correct
    // AbstractBase<Integer> 中的 <Integer> 不能省略
    static class Sample1 extends AbstractBase<Integer> {

        Sample1() {
            super(10);
        }

    }

    // incorrect
//    static class Sample2 extends AbstractBase<>{
//        Sample2() {
//            super(10);
//        }
//    }
    static class Sample3<T> {

        Sample3(T value) {
        }

    }

    @Test
    @DisplayName("继承的 Constructor 不会推断类型")
    void test300() {
        AbstractBase<Integer> sample1 = new Sample1();
        // correct
        Sample3<Integer> sample3 = new Sample3<>(10);
    }

}

File: kotlin/src/test/kotlin/cn/staynoob/trap/kotlin/basic/TypeInferenceSpec.kt
package cn.staynoob.trap.kotlin.basic

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import kotlin.reflect.full.createInstance

class TypeInferenceSpec {

    interface Producer<T> {
        fun produce(): T
    }

    class StringProducer : Producer<String> {
        override fun produce(): String {
            return "42"
        }
    }

    inline fun <reified S : Producer<T>, T> underscoreTypeArgExample(): T {
        return S::class.createInstance().produce()
    }

    @Test
    @DisplayName("underscore type argument")
    fun test100() {
        val a = underscoreTypeArgExample<StringProducer, _>()
        assertThat(a).isEqualTo("42")
    }

    @Test
    @DisplayName("继承调用父类构造函数时，即使可以推断类型，也需要显式声名")
    fun test1000() {
        abstract class Base<T>(value: T)

        // correct
        class Sample1 : Base<Int>(10)

        // incorrect
//        class Sample2: Base(10)
//        class Sample2: Base<>(10)
//        class Sample2: Base<_>(10)
        // correct
        // 直接调用父类构造函数，可以省略类型参数
        class Sample3<T>(value: T)

        val sample3 = Sample3(10)
    }

}
